### DISCUSSION TITLE
shadowhook_hook_func_addr 中绝对函数地址如何获取呢？如果无法使用shadowhook_dlsym_symtab获取到地址

### DISCUSSION THREAD

**Discussion by uchiaitachipro** [bytedance/android-inline-hook#30] [Q&A]
例如我想hook libjavacrypto.so中sub_1E7C0 函数，它没有导出。
<img width="778" alt="image" src="https://user-images.githubusercontent.com/10720848/206884169-9802f117-7bfb-4d58-bfe3-04b8d03d6fb0.png">

**Comment by caikelun** [comment]
常见的方法有：

1. 如果对于固定版本的so库，函数的offset是固定的，所以绝对地址 = so的load bias + 函数相对于so的offset。
2. 如果调用指令所在的函数（如你截图中的JNI_OnLoad）结构比较简单，可以解析函数中的指令序列特征，找到调用对应的BL指令，解析BL指令获取相对地址，再加上当前pc就是目标函数的绝对地址。

load bias可以用 `dl_iterate_phdr()` 来获取。
